packages/camera/camera_avfoundation/ios/camera_avfoundation/Sources/camera_avfoundation/FLTCam.m

```diff
@@ -211,20 +211,50 @@ - (instancetype)initWithMediaSettings:(FCPPlatformMediaSettings *)mediaSettings
   [_motionManager startAccelerometerUpdates];
 
   if (_mediaSettings.framesPerSecond) {
+    [_mediaSettingsAVWrapper beginConfigurationForSession:_videoCaptureSession];
+
+    // Possible values for presets are hard-coded in FLT interface having
+    // corresponding AVCaptureSessionPreset counterparts.
+    // If _resolutionPreset is not supported by camera there is
+    // fallback to lower resolution presets.
+    // If none can be selected there is error condition.
+    if (![self setCaptureSessionPreset:_mediaSettings.resolutionPreset withError:error]) {
+      [_videoCaptureSession commitConfiguration];
+      return nil;
+    }
+
     // The frame rate can be changed only on a locked for configuration device.
     if ([mediaSettingsAVWrapper lockDevice:_captureDevice error:error]) {
-      [_mediaSettingsAVWrapper beginConfigurationForSession:_videoCaptureSession];
-
-      // Possible values for presets are hard-coded in FLT interface having
-      // corresponding AVCaptureSessionPreset counterparts.
-      // If _resolutionPreset is not supported by camera there is
-      // fallback to lower resolution presets.
-      // If none can be selected there is error condition.
-      if (![self setCaptureSessionPreset:_mediaSettings.resolutionPreset withError:error]) {
-        [_videoCaptureSession commitConfiguration];
-        [_captureDevice unlockForConfiguration];
-        return nil;
+      // find the format which frame rate ranges are closest to the wanted frame rate
+      CMVideoDimensions targetRes = self.videoDimensionsForFormat(_captureDevice.activeFormat);
+      double targetFrameRate = _mediaSettings.framesPerSecond.doubleValue;
+      FourCharCode preferredSubType =
+          CMFormatDescriptionGetMediaSubType(_captureDevice.activeFormat.formatDescription);
+      AVCaptureDeviceFormat *bestFormat = _captureDevice.activeFormat;
+      double bestFrameRate = [self frameRateForFormat:bestFormat closestTo:targetFrameRate];
+      double minDistance = fabs(bestFrameRate - targetFrameRate);
+      int bestSubTypeScore = 1;
+      for (AVCaptureDeviceFormat *format in _captureDevice.formats) {
+        CMVideoDimensions res = self.videoDimensionsForFormat(format);
+        if (res.width != targetRes.width || res.height != targetRes.height) {
+          continue;
+        }
+        double frameRate = [self frameRateForFormat:format closestTo:targetFrameRate];
+        double distance = fabs(frameRate - targetFrameRate);
+        FourCharCode subType = CMFormatDescriptionGetMediaSubType(format.formatDescription);
+        int subTypeScore = subType == preferredSubType ? 1 : 0;
+        if (distance < minDistance ||
+            (distance == minDistance && subTypeScore > bestSubTypeScore)) {
```

Comment:
I'm confused by this condition. `bestSubTypeScore` is initialized to 1, and `subTypeScore` is always either 0 or 1. How could this last piece of the condition ever be true?

I'm not clear on what this score even is; this whole algorithm needs some explanatory comment at the beginning.

Reply:
if `distance < minDistance` then `bestSubTypeScore` can be overwritten by 0. This line explains it: `subTypeScore = subType == preferredSubType ? 1 : 0`, the score of the subtype is 1 when it is the preferred subtype and 0 otherwise. This is how to choose by minimum distance and then from formats which have equal distance choose that one with preferred subtype, it gives to subtypes some order. Instead of `subTypeScore > bestSubTypeScore` there could be something like `bestSubType != preferredSubType && subType == preferredSubType` but that makes it less clear as it deviates from simple way how to order something by multiple properties `currentA < otherA || (currentA == otherA && currentB < otherB)`.

Reply:
> but that makes it less clear

I strongly disagree that abstracting the thing that matters—whether the current type is a preferred type—behind a completely arbitrary numeric value makes this code easier to understand. The code you have there that you described as "less clear" makes sense to me when I read it, whereas I did not understand what this score was, or why something that could only have two possible values was being called a score in the first place.